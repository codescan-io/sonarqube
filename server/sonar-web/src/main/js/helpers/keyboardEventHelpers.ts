/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
export function isShortcut(event: KeyboardEvent): boolean {
  return event.ctrlKey || event.metaKey;
}

export function isTextarea(
  event: KeyboardEvent
): event is KeyboardEvent & { target: HTMLTextAreaElement } {
  return event.target instanceof HTMLTextAreaElement;
}

export function isInput(
  event: KeyboardEvent
): event is KeyboardEvent & { target: HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement } {
  return (
    event.target instanceof HTMLInputElement ||
    event.target instanceof HTMLSelectElement ||
    event.target instanceof HTMLTextAreaElement
  );
}

/*
 * Due to React 16 event delegation, stopPropagation called within react-day-picker is NOT preventing other event handlers from being called.
 * As a temporary workaround, we detect this special case using this utility function.
 * This utility function can be removed once we upgrade to React 17, since although there is still event delegation,
 * it is delegated up to the React root, which will stop propagation before it reaches document event handlers.
 */
export function isDatePicker(event: KeyboardEvent): boolean {
  return event.target instanceof Element && event.target.matches('.rdp-day');
}
