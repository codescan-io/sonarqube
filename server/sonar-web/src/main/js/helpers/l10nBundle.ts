/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { fetchL10nBundle } from '../api/l10n';
import { L10nBundle, L10nBundleRequestParams } from '../types/l10nBundle';
import { toNotSoISOString } from './dates';

const DEFAULT_LOCALE = 'en';
const DEFAULT_MESSAGES = {
  // eslint-disable-next-line camelcase
  default_error_message: 'The request cannot be processed. Try again later.',
};

export function getMessages() {
  return getL10nBundleFromCache().messages ?? DEFAULT_MESSAGES;
}

export function getCurrentLocale() {
  return getL10nBundleFromCache().locale;
}

export function getCurrentL10nBundle() {
  return getL10nBundleFromCache();
}

export async function loadL10nBundle() {
  const browserLocale = getPreferredLanguage();
  const cachedBundle = getL10nBundleFromCache();

  const params: L10nBundleRequestParams = {};

  if (browserLocale) {
    params.locale = browserLocale;

    if (
      cachedBundle.locale &&
      browserLocale.startsWith(cachedBundle.locale) &&
      cachedBundle.timestamp &&
      cachedBundle.messages
    ) {
      params.ts = cachedBundle.timestamp;
    }
  }

  const { effectiveLocale, messages } = await fetchL10nBundle(params).catch((response) => {
    if (response && response.status === 304) {
      return {
        effectiveLocale: cachedBundle.locale || browserLocale || DEFAULT_LOCALE,
        messages: cachedBundle.messages ?? {},
      };
    }
    throw new Error(`Unexpected status code: ${response.status}`);
  });

  const bundle = {
    timestamp: toNotSoISOString(new Date()),
    locale: effectiveLocale,
    messages,
  };

  persistL10nBundleInCache(bundle);

  return bundle;
}

function getPreferredLanguage() {
  return window.navigator.languages ? window.navigator.languages[0] : window.navigator.language;
}

function getL10nBundleFromCache(): L10nBundle {
  return (window as unknown as any).sonarQubeL10nBundle ?? {};
}

function persistL10nBundleInCache(bundle: L10nBundle) {
  (window as unknown as any).sonarQubeL10nBundle = bundle;
}
