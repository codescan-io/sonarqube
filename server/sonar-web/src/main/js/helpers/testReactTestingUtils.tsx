/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { render, RenderResult } from '@testing-library/react';
import * as React from 'react';
import { HelmetProvider } from 'react-helmet-async';
import { IntlProvider } from 'react-intl';
import { MemoryRouter, Outlet, parsePath, Route, Routes } from 'react-router-dom';
import AdminContext from '../app/components/AdminContext';
import AppStateContextProvider from '../app/components/app-state/AppStateContextProvider';
import { AvailableFeaturesContext } from '../app/components/available-features/AvailableFeaturesContext';
import { ComponentContext } from '../app/components/componentContext/ComponentContext';
import CurrentUserContextProvider from '../app/components/current-user/CurrentUserContextProvider';
import GlobalMessagesContainer from '../app/components/GlobalMessagesContainer';
import IndexationContextProvider from '../app/components/indexation/IndexationContextProvider';
import { LanguagesContext } from '../app/components/languages/LanguagesContext';
import { MetricsContext } from '../app/components/metrics/MetricsContext';
import { useLocation } from '../components/hoc/withRouter';
import { AppState } from '../types/appstate';
import { ComponentContextShape } from '../types/component';
import { Feature } from '../types/features';
import { Dict, Extension, Languages, Metric, SysStatus } from '../types/types';
import { CurrentUser } from '../types/users';
import { DEFAULT_METRICS } from './mocks/metrics';
import { mockAppState, mockCurrentUser } from './testMocks';

export interface RenderContext {
  metrics?: Dict<Metric>;
  appState?: AppState;
  languages?: Languages;
  currentUser?: CurrentUser;
  navigateTo?: string;
  featureList?: Feature[];
}

export function renderAppWithAdminContext(
  indexPath: string,
  routes: () => JSX.Element,
  context: RenderContext = {},
  overrides: { systemStatus?: SysStatus; adminPages?: Extension[] } = {}
): RenderResult {
  function MockAdminContainer() {
    return (
      <AdminContext.Provider
        value={{
          fetchSystemStatus: () => {
            /*noop*/
          },
          fetchPendingPlugins: () => {
            /*noop*/
          },
          pendingPlugins: { installing: [], removing: [], updating: [] },
          systemStatus: overrides.systemStatus ?? 'UP',
        }}
      >
        <Outlet
          context={{
            adminPages: overrides.adminPages ?? [],
          }}
        />
      </AdminContext.Provider>
    );
  }

  return renderRoutedApp(
    <Route element={<MockAdminContainer />} path="admin">
      {routes()}
    </Route>,
    indexPath,
    context
  );
}

export function renderComponent(component: React.ReactElement, pathname = '/') {
  function Wrapper({ children }: { children: React.ReactElement }) {
    return (
      <IntlProvider defaultLocale="en" locale="en">
        <MemoryRouter initialEntries={[pathname]}>
          <Routes>
            <Route path="*" element={children} />
          </Routes>
        </MemoryRouter>
      </IntlProvider>
    );
  }

  return render(component, { wrapper: Wrapper });
}

export function renderAppWithComponentContext(
  indexPath: string,
  routes: () => JSX.Element,
  context: RenderContext = {},
  componentContext?: Partial<ComponentContextShape>
) {
  function MockComponentContainer() {
    return (
      <ComponentContext.Provider
        value={{
          branchLikes: [],
          onBranchesChange: jest.fn(),
          onComponentChange: jest.fn(),
          ...componentContext,
        }}
      >
        <Outlet />
      </ComponentContext.Provider>
    );
  }

  return renderRoutedApp(
    <Route element={<MockComponentContainer />}>{routes()}</Route>,
    indexPath,
    context
  );
}

export function renderApp(
  indexPath: string,
  component: JSX.Element,
  context: RenderContext = {}
): RenderResult {
  return renderRoutedApp(<Route path={indexPath} element={component} />, indexPath, context);
}

export function renderAppRoutes(
  indexPath: string,
  routes: () => JSX.Element,
  context?: RenderContext
): RenderResult {
  return renderRoutedApp(routes(), indexPath, context);
}

export function CatchAll() {
  const location = useLocation();

  return <div>{`${location.pathname}${location.search}`}</div>;
}

function renderRoutedApp(
  children: React.ReactElement,
  indexPath: string,
  {
    currentUser = mockCurrentUser(),
    navigateTo = indexPath,
    metrics = DEFAULT_METRICS,
    appState = mockAppState(),
    featureList = [],
    languages = {},
  }: RenderContext = {}
): RenderResult {
  const path = parsePath(navigateTo);
  path.pathname = `/${path.pathname}`;

  return render(
    <HelmetProvider context={{}}>
      <IntlProvider defaultLocale="en" locale="en">
        <MetricsContext.Provider value={metrics}>
          <LanguagesContext.Provider value={languages}>
            <AvailableFeaturesContext.Provider value={featureList}>
              <CurrentUserContextProvider currentUser={currentUser}>
                <AppStateContextProvider appState={appState}>
                  <IndexationContextProvider>
                    <GlobalMessagesContainer />
                    <MemoryRouter initialEntries={[path]}>
                      <Routes>
                        {children}
                        <Route path="*" element={<CatchAll />} />
                      </Routes>
                    </MemoryRouter>
                  </IndexationContextProvider>
                </AppStateContextProvider>
              </CurrentUserContextProvider>
            </AvailableFeaturesContext.Provider>
          </LanguagesContext.Provider>
        </MetricsContext.Provider>
      </IntlProvider>
    </HelmetProvider>
  );
}
