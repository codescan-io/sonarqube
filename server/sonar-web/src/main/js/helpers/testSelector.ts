/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import {
  BoundFunction,
  GetByBoundAttribute,
  GetByRole,
  GetByText,
  screen,
  waitForOptions,
  within,
} from '@testing-library/react';

function maybeScreen(container?: HTMLElement) {
  return container ? within(container) : screen;
}

export interface ReactTestingQuery {
  find<T extends HTMLElement = HTMLElement>(
    container?: HTMLElement,
    waitForOptions?: waitForOptions
  ): Promise<T>;
  findAll<T extends HTMLElement = HTMLElement>(
    container?: HTMLElement,
    waitForOptions?: waitForOptions
  ): Promise<T[]>;
  get<T extends HTMLElement = HTMLElement>(container?: HTMLElement): T;
  getAll<T extends HTMLElement = HTMLElement>(container?: HTMLElement): T[];
  query<T extends HTMLElement = HTMLElement>(container?: HTMLElement): T | null;
  queryAll<T extends HTMLElement = HTMLElement>(container?: HTMLElement): T[] | null;

  byText(...args: Parameters<BoundFunction<GetByText>>): ReactTestingQuery;
  byRole(...args: Parameters<BoundFunction<GetByRole>>): ReactTestingQuery;
  byPlaceholderText(...args: Parameters<BoundFunction<GetByBoundAttribute>>): ReactTestingQuery;
  byLabelText(...args: Parameters<BoundFunction<GetByText>>): ReactTestingQuery;
  byTestId(...args: Parameters<BoundFunction<GetByBoundAttribute>>): ReactTestingQuery;
  byDisplayValue(...args: Parameters<BoundFunction<GetByBoundAttribute>>): ReactTestingQuery;
}

abstract class ChainingQuery implements ReactTestingQuery {
  abstract find<T extends HTMLElement = HTMLElement>(
    container?: HTMLElement,
    waitForOptions?: waitForOptions
  ): Promise<T>;

  abstract findAll<T extends HTMLElement = HTMLElement>(
    container?: HTMLElement,
    waitForOptions?: waitForOptions
  ): Promise<T[]>;

  abstract get<T extends HTMLElement = HTMLElement>(container?: HTMLElement): T;

  abstract getAll<T extends HTMLElement = HTMLElement>(container?: HTMLElement): T[];

  abstract query<T extends HTMLElement = HTMLElement>(container?: HTMLElement): T | null;

  abstract queryAll<T extends HTMLElement = HTMLElement>(container?: HTMLElement): T[] | null;

  byText(...args: Parameters<BoundFunction<GetByText>>): ReactTestingQuery {
    return new ChainDispatch(this, new DispatchByText(args));
  }

  byRole(...args: Parameters<BoundFunction<GetByRole>>): ReactTestingQuery {
    return new ChainDispatch(this, new DispatchByRole(args));
  }

  byPlaceholderText(...args: Parameters<BoundFunction<GetByBoundAttribute>>): ReactTestingQuery {
    return new ChainDispatch(this, new DispatchByPlaceholderText(args));
  }

  byLabelText(...args: Parameters<BoundFunction<GetByText>>): ReactTestingQuery {
    return new ChainDispatch(this, new DispatchByLabelText(args));
  }

  byTestId(...args: Parameters<BoundFunction<GetByBoundAttribute>>): ReactTestingQuery {
    return new ChainDispatch(this, new DispatchByTestId(args));
  }

  byDisplayValue(...args: Parameters<BoundFunction<GetByBoundAttribute>>): ReactTestingQuery {
    return new ChainDispatch(this, new DispatchByDisplayValue(args));
  }
}

class ChainDispatch extends ChainingQuery {
  insideQuery: ReactTestingQuery;
  elementQuery: ReactTestingQuery;

  constructor(insideQuery: ReactTestingQuery, elementQuery: ReactTestingQuery) {
    super();
    this.insideQuery = insideQuery;
    this.elementQuery = elementQuery;
  }

  async find<T extends HTMLElement = HTMLElement>(
    container?: HTMLElement,
    waitForOptions?: waitForOptions
  ) {
    return this.elementQuery.get<T>(await this.insideQuery.find(container, waitForOptions));
  }

  async findAll<T extends HTMLElement = HTMLElement>(
    container?: HTMLElement,
    waitForOptions?: waitForOptions
  ) {
    return this.elementQuery.getAll<T>(await this.insideQuery.find(container, waitForOptions));
  }

  get<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return this.elementQuery.get<T>(this.insideQuery.get(container));
  }

  getAll<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return this.elementQuery.getAll<T>(this.insideQuery.get(container));
  }

  query<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return this.elementQuery.query<T>(this.insideQuery.get(container));
  }

  queryAll<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return this.elementQuery.queryAll<T>(this.insideQuery.get(container));
  }
}

class DispatchByText extends ChainingQuery {
  readonly args: Parameters<BoundFunction<GetByText>>;

  constructor(args: Parameters<BoundFunction<GetByText>>) {
    super();
    this.args = args;
  }

  find<T extends HTMLElement = HTMLElement>(
    container?: HTMLElement,
    waitForOptions?: waitForOptions
  ) {
    return maybeScreen(container).findByText<T>(...this.args, waitForOptions);
  }

  findAll<T extends HTMLElement = HTMLElement>(
    container?: HTMLElement,
    waitForOptions?: waitForOptions
  ) {
    return maybeScreen(container).findAllByText<T>(...this.args, waitForOptions);
  }

  get<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).getByText<T>(...this.args);
  }

  getAll<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).getAllByText<T>(...this.args);
  }

  query<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).queryByText<T>(...this.args);
  }

  queryAll<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).queryAllByText<T>(...this.args);
  }
}

class DispatchByLabelText extends ChainingQuery {
  readonly args: Parameters<BoundFunction<GetByText>>;

  constructor(args: Parameters<BoundFunction<GetByText>>) {
    super();
    this.args = args;
  }

  find<T extends HTMLElement = HTMLElement>(
    container?: HTMLElement,
    waitForOptions?: waitForOptions
  ) {
    return maybeScreen(container).findByLabelText<T>(...this.args, waitForOptions);
  }

  findAll<T extends HTMLElement = HTMLElement>(
    container?: HTMLElement,
    waitForOptions?: waitForOptions
  ) {
    return maybeScreen(container).findAllByLabelText<T>(...this.args, waitForOptions);
  }

  get<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).getByLabelText<T>(...this.args);
  }

  getAll<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).getAllByLabelText<T>(...this.args);
  }

  query<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).queryByLabelText<T>(...this.args);
  }

  queryAll<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).queryAllByLabelText<T>(...this.args);
  }
}

class DispatchByRole extends ChainingQuery {
  readonly args: Parameters<BoundFunction<GetByRole>>;

  constructor(args: Parameters<BoundFunction<GetByRole>>) {
    super();
    this.args = args;
  }

  find<T extends HTMLElement = HTMLElement>(
    container?: HTMLElement,
    waitForOptions?: waitForOptions
  ) {
    return maybeScreen(container).findByRole<T>(...this.args, waitForOptions);
  }

  findAll<T extends HTMLElement = HTMLElement>(
    container?: HTMLElement,
    waitForOptions?: waitForOptions
  ) {
    return maybeScreen(container).findAllByRole<T>(...this.args, waitForOptions);
  }

  get<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).getByRole<T>(...this.args);
  }

  getAll<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).getAllByRole<T>(...this.args);
  }

  query<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).queryByRole<T>(...this.args);
  }

  queryAll<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).queryAllByRole<T>(...this.args);
  }
}

class DispatchByTestId extends ChainingQuery {
  readonly args: Parameters<BoundFunction<GetByBoundAttribute>>;

  constructor(args: Parameters<BoundFunction<GetByBoundAttribute>>) {
    super();
    this.args = args;
  }

  find<T extends HTMLElement = HTMLElement>(
    container?: HTMLElement,
    waitForOptions?: waitForOptions
  ) {
    return maybeScreen(container).findByTestId<T>(...this.args, waitForOptions);
  }

  findAll<T extends HTMLElement = HTMLElement>(
    container?: HTMLElement,
    waitForOptions?: waitForOptions
  ) {
    return maybeScreen(container).findAllByTestId<T>(...this.args, waitForOptions);
  }

  get<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).getByTestId<T>(...this.args);
  }

  getAll<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).getAllByTestId<T>(...this.args);
  }

  query<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).queryByTestId<T>(...this.args);
  }

  queryAll<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).queryAllByTestId<T>(...this.args);
  }
}

class DispatchByDisplayValue extends ChainingQuery {
  readonly args: Parameters<BoundFunction<GetByBoundAttribute>>;

  constructor(args: Parameters<BoundFunction<GetByBoundAttribute>>) {
    super();
    this.args = args;
  }

  find<T extends HTMLElement = HTMLElement>(
    container?: HTMLElement,
    waitForOptions?: waitForOptions
  ) {
    return maybeScreen(container).findByDisplayValue<T>(...this.args, waitForOptions);
  }

  findAll<T extends HTMLElement = HTMLElement>(
    container?: HTMLElement,
    waitForOptions?: waitForOptions
  ) {
    return maybeScreen(container).findAllByDisplayValue<T>(...this.args, waitForOptions);
  }

  get<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).getByDisplayValue<T>(...this.args);
  }

  getAll<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).getAllByDisplayValue<T>(...this.args);
  }

  query<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).queryByDisplayValue<T>(...this.args);
  }

  queryAll<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).queryAllByDisplayValue<T>(...this.args);
  }
}

class DispatchByPlaceholderText extends ChainingQuery {
  readonly args: Parameters<BoundFunction<GetByBoundAttribute>>;

  constructor(args: Parameters<BoundFunction<GetByBoundAttribute>>) {
    super();
    this.args = args;
  }

  find<T extends HTMLElement = HTMLElement>(
    container?: HTMLElement,
    waitForOptions?: waitForOptions
  ) {
    return maybeScreen(container).findByPlaceholderText<T>(...this.args, waitForOptions);
  }

  findAll<T extends HTMLElement = HTMLElement>(
    container?: HTMLElement,
    waitForOptions?: waitForOptions
  ) {
    return maybeScreen(container).findAllByPlaceholderText<T>(...this.args, waitForOptions);
  }

  get<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).getByPlaceholderText<T>(...this.args);
  }

  getAll<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).getAllByPlaceholderText<T>(...this.args);
  }

  query<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).queryByPlaceholderText<T>(...this.args);
  }

  queryAll<T extends HTMLElement = HTMLElement>(container?: HTMLElement) {
    return maybeScreen(container).queryAllByPlaceholderText<T>(...this.args);
  }
}

export function byText(...args: Parameters<BoundFunction<GetByText>>) {
  return new DispatchByText(args);
}

export function byRole(...args: Parameters<BoundFunction<GetByRole>>) {
  return new DispatchByRole(args);
}

export function byPlaceholderText(...args: Parameters<BoundFunction<GetByBoundAttribute>>) {
  return new DispatchByPlaceholderText(args);
}

export function byLabelText(...args: Parameters<BoundFunction<GetByText>>) {
  return new DispatchByLabelText(args);
}

export function byTestId(...args: Parameters<BoundFunction<GetByBoundAttribute>>) {
  return new DispatchByTestId(args);
}

export function byDisplayValue(...args: Parameters<BoundFunction<GetByBoundAttribute>>) {
  return new DispatchByDisplayValue(args);
}
