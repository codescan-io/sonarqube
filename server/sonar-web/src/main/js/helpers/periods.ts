/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { parseDate } from '../helpers/dates';
import { translate, translateWithParameters } from '../helpers/l10n';
import { ApplicationPeriod } from '../types/application';
import { NewCodePeriod, NewCodePeriodSettingType, Period } from '../types/types';

export function getPeriodLabel(
  period: Period | undefined,
  dateFormatter: (date: string) => string
) {
  if (!period) {
    return undefined;
  }

  let parameter = period.modeParam || period.parameter || '';

  switch (period.mode) {
    case NewCodePeriodSettingType.SPECIFIC_ANALYSIS:
      parameter = dateFormatter(period.date);
      break;
    case NewCodePeriodSettingType.PREVIOUS_VERSION:
      parameter = parameter || dateFormatter(period.date);
      break;
    /*
     * Handle legacy period modes, that predate MMF-1579
     */
    case 'previous_version':
      if (!parameter) {
        return translate('overview.period.previous_version_only_date');
      }
      break;
    case 'date':
      parameter = parameter && dateFormatter(parameter);
      break;
    case 'manual_baseline':
      parameter = parameter || dateFormatter(period.date);
      break;
    default: // No change in the parameter
  }

  return translateWithParameters(`overview.period.${period.mode.toLowerCase()}`, parameter);
}

export function getPeriodDate(period?: { date?: string }): Date | undefined {
  return period && period.date ? parseDate(period.date) : undefined;
}

export function isApplicationPeriod(
  period: Period | ApplicationPeriod
): period is ApplicationPeriod {
  return (period as ApplicationPeriod).project !== undefined;
}

const MIN_NUMBER_OF_DAYS = 1;
const MAX_NUMBER_OF_DAYS = 90;

export function isNewCodeDefinitionCompliant(newCodePeriod: NewCodePeriod) {
  switch (newCodePeriod.type) {
    case NewCodePeriodSettingType.NUMBER_OF_DAYS:
      return (
        newCodePeriod.value !== undefined &&
        MIN_NUMBER_OF_DAYS <= +newCodePeriod.value &&
        +newCodePeriod.value <= MAX_NUMBER_OF_DAYS
      );
    default:
      return true;
  }
}
