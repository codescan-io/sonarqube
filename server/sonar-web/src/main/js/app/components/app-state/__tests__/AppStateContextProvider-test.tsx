/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { mount } from 'enzyme';
import * as React from 'react';
import { mockAppState } from '../../../../helpers/testMocks';
import { waitAndUpdate } from '../../../../helpers/testUtils';
import AppStateContextProvider, { AppStateContextProviderProps } from '../AppStateContextProvider';

it('should set value correctly', async () => {
  const appState = mockAppState({ settings: { 'sonar.lf.logoUrl': 'whatevs/' } });
  const wrapper = render({
    appState,
  });
  await waitAndUpdate(wrapper);

  expect(wrapper).toMatchSnapshot();
});

function render(override?: Partial<AppStateContextProviderProps>) {
  return mount(<AppStateContextProvider appState={mockAppState()} {...override} />);
}
