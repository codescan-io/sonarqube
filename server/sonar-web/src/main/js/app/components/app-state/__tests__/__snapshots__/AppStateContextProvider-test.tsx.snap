// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should set value correctly 1`] = `
<AppStateContextProvider
  appState={
    {
      "edition": "community",
      "productionDatabase": true,
      "qualifiers": [
        "TRK",
      ],
      "settings": {
        "sonar.lf.logoUrl": "whatevs/",
      },
      "version": "1.0",
    }
  }
/>
`;
